client/src/components/Navbar.js

```diff
@@ -19,35 +91,31 @@ const NavBar = props => {
         </Navbar.Brand>
         <Navbar.Toggle />
       </Navbar.Header>
-      <Navbar.Collapse>
-        <Nav pullRight>
-          <NavItem eventKey={1}>
-            <Link to="/denver-learn-more">Why</Link>
-          </NavItem>
-          <NavDropdown id="tools-dropdown" eventKey={2} title="Tools" href="#">
-            <MenuItem eventKey={2.1}>
-              <Link to="/manager-resources">Property Manager Resources</Link>
-            </MenuItem>
-            <MenuItem eventKey={2.2}>
-              <Link to="/tips-for-requesting">Tips for Requesting</Link>
-            </MenuItem>
-          </NavDropdown>
-          <NavItem eventKey={3}>
-            <Link to="">My Campaign</Link>
-          </NavItem>
-          <NavItem eventKey={4}>
-            <Link to="/who-are-we">Who Are We</Link>
-          </NavItem>
```

Comment:
Keep the common jsx here, call to UserNavItem (which could be either a function or a component) to add whatever the login/logout logic specifies.

Reply:
So I started with this idea but couldn't implement it for the following: 

UserIsNotLoggedIn would work no problem as it returns one NavLink. 

UserIsLoggedIn needs to return two NavLinks (My Campaign and Logout). Since we are still in React 15 you can't return two top level elements and we don't have access to Fragments. I tried wrapping this in different elements <div> etc but it negatively affects the render of the Navbar. Do you know a way to implement this? I agree with you and support the change if you have the magic bullet.

Reply:
ah, good observation.  if you cannot return multiple and wrapping in a div or some other element throws off the css our best bet is to separate the two different pieces into their separate logical returns.  


I was able to quickly pull something together that works, definitely not polished code though.
(the UserNavItem were each reduced to the smallest necessary size,  these components might need renaming now that they are performing different roles).

What are your thoughts on structuring the code like this?

const isLoggedIn =
    props.auth && (props.auth.googleID || props.auth.facebookID);
  return (
    <Navbar bsStyle="remove-default" collapseOnSelect fluid>
      <Navbar.Header>
        <Navbar.Brand>
          <Link to="/">{homeText}</Link>
        </Navbar.Brand>
        <Navbar.Toggle />
      </Navbar.Header>
      <Navbar.Collapse>
        <Nav pullRight>
          <NavItem eventKey={1}>
            <Link to="/denver-learn-more">Why</Link>
          </NavItem>
          <NavDropdown id="tools-dropdown" eventKey={2} title="Tools">
            <MenuItem eventKey={2.1}>
              <Link to="/manager-resources">Property Manager Resources</Link>
            </MenuItem>
            <MenuItem eventKey={2.2}>
              <Link to="/tips-for-requesting">Tips for Requesting</Link>
            </MenuItem>
          </NavDropdown>
          <NavItem eventKey={3}>
            <Link to="/who-are-we">Who Are We</Link>
          </NavItem>
          {isLoggedIn && (
            <NavItem eventKey={4}>
              <Link to="">My Campaign</Link>
            </NavItem>
          )}
          <UserNavItem {...props} />
        </Nav>
      </Navbar.Collapse>
    </Navbar>
  );
};

Reply:
haha yea that's a mess. Let me take a swing at it but basically have a UserAuthNavItem that either provides login/logout logic and a MyCampiagnNavItem that either renders or not based on Auth. I think that should solve it. And I think that's less complicated?

Reply:
yes. however i would not have MyCampiagnNavItem decide whether or not to render itself,  the object rendering it should be the one who contains the logic on whether it shows up or not.